contracts/ronin/gateway/BridgeReward.sol

```diff
@@ -0,0 +1,256 @@
+
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  /**
+   * @dev Returns whether should share the reward equally, in case of bridge tracking returns
+   * informed data or there is no ballot in a day.
+   *
+   * Emit a {BridgeTrackingIncorrectlyResponded} event when in case of incorrect data.
+   */
+  function _isSharingRewardEqually(
+    uint256 totalBallots,
+    uint256 totalVotes,
+    uint256[] memory ballots
+  ) internal returns (bool shareEqually) {
+    bool valid = _isValidBridgeTrackingResponse(totalBallots, totalVotes, ballots);
+    if (!valid) {
+      emit BridgeTrackingIncorrectlyResponded();
+    }
+
+    return !valid || totalBallots == 0;
+  }
+
+  function _calcRewardAndCheckSlashedStatus(
+    bool isSharingRewardEqually,
+    uint256 numBridgeOperators,
+    uint256 rewardPerPeriod,
+    uint256 ballot,
+    uint256 totalBallots,
+    uint256 period,
+    uint256 slashUntilPeriod
+  ) internal pure returns (uint256 reward, bool isSlashed) {
+    isSlashed = _isSlashedThisPeriod(period, slashUntilPeriod);
+    reward = _calcReward(isSharingRewardEqually, numBridgeOperators, rewardPerPeriod, ballot, totalBallots);
+  }
+
+  function _isSlashedThisPeriod(uint256 period, uint256 slashDuration) internal pure returns (bool) {
+    return period <= slashDuration;
+  }
+
+  function _calcReward(
+    bool isSharingRewardEqually,
+    uint256 numBridgeOperators,
+    uint256 rewardPerPeriod,
+    uint256 ballot,
+    uint256 totalBallots
+  ) internal pure returns (uint256 reward) {
+    // Shares equally in case the bridge has nothing to vote or bridge tracking response is incorrect
+    // Else shares the bridge operators reward proportionally
+    reward = isSharingRewardEqually ? rewardPerPeriod / numBridgeOperators : (rewardPerPeriod * ballot) / totalBallots;
+  }
+
+  function _isValidBridgeTrackingResponse(
+    uint256 totalBallots,
+    uint256 totalVotes,
+    uint256[] memory ballots
+  ) internal pure returns (bool valid) {
+    valid = true;
+    uint256 sumBallots;
+    for (uint _i; _i < ballots.length; _i++) {
+      if (ballots[_i] > totalVotes) {
+        valid = false;
+        break;
+      }
+      sumBallots += ballots[_i];
+    }
+    valid = valid && (sumBallots <= totalBallots);
+  }
+
+  /**
+   * @dev Transfer `reward` to a `operator` or only emit event based on the operator `slashed` status.
+   */
+  function _updateRewardAndTransfer(address operator, uint256 reward, bool isSlashed) private {
+    BridgeRewardInfo storage _iRewardInfo = _rewardInfo[operator];
+    if (isSlashed) {
+      _iRewardInfo.slashed += reward;
+      emit BridgeRewardSlashed(operator, reward);
+    } else {
+      _iRewardInfo.claimed += reward;
+      if (_unsafeSendRONLimitGas({ recipient: payable(operator), amount: reward, gas: 0 })) {
+        emit BridgeRewardScattered(operator, reward);
+      } else {
+        emit BridgeRewardScatterFailed(operator, reward);
```

Comment:
These 3 events should emitted with the period

Reply:
updated.